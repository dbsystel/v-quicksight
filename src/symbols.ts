export const EmbeddingContextInjectionKey = Symbol(
  "The key to be used with inject when accessing the created embeddingContext for usage with the amazon-quicksight-embedding-sdk",
)
