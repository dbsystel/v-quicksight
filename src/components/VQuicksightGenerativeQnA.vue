<script setup lang="ts">
import type { VQuicksightGenerativeQnAContentOptions, VQuicksightFrameOptions } from '../types'
import type {
  GenerativeQnAContentOptions,
  GenerativeQnAExperience,
  EmbeddingContext,
  ExperienceFrameMetadata,
  FrameOptions,
  EmbeddingEvents
} from 'amazon-quicksight-embedding-sdk'
import { nanoid } from 'nanoid'
import type { Ref } from 'vue'
import { computed, inject, ref, watch } from 'vue'
import { EmbeddingContextInjectionKey } from '../symbols'

const props = withDefaults(
  defineProps<
    VQuicksightFrameOptions &
      VQuicksightGenerativeQnAContentOptions & {
        container?: string | HTMLElement // override from FrameOptions and make it optional
        id?: string
        question?: string
      }
  >(),
  {
    width: '100%',
    height: '100%',
    withIframePlaceholder: false,
    showTopicName: true,
    showPinboard: true,
    allowTopicSelection: true,
    allowFullscreen: true
  }
)

const emit = defineEmits<{
  (e: 'change', data: { changeEvent: EmbeddingEvents; metadata?: ExperienceFrameMetadata }): void
  (
    e: 'message',
    data: { messageEvent: EmbeddingEvents; experienceMetadata?: ExperienceFrameMetadata }
  ): void
}>()

const embeddingContext = inject<Ref<EmbeddingContext>>(EmbeddingContextInjectionKey)

const qnaFrame = ref<GenerativeQnAExperience>()

const containerId = computed(() => props.id || `v-quicksight-generative-qna-${nanoid(6)}`)
const frameOptions = computed<FrameOptions>(() => {
  return {
    url: props.url,
    container: props.container || `#${containerId.value}`,
    height: props.height,
    width: props.width,
    className: props.className,
    withIframePlaceholder: props.withIframePlaceholder,
    onChange: (changeEvent, metadata) => {
      emit('change', { changeEvent, metadata })
    }
  }
})
const contentOptions = computed<GenerativeQnAContentOptions>(() => {
  return {
    showTopicName: props.showTopicName,
    showPinboard: props.showPinboard,
    allowTopicSelection: props.allowTopicSelection,
    allowFullscreen: props.allowFullscreen,
    searchPlaceholderText: props.searchPlaceholderText,
    panelOptions: props.panelOptions,
    themeOptions: props.themeOptions,
    onMessage: async (messageEvent, experienceMetadata) => {
      emit('message', { messageEvent, experienceMetadata })
    }
  }
})

async function embedQnA(ctx: EmbeddingContext) {
  qnaFrame.value = await ctx.embedGenerativeQnA(frameOptions.value, contentOptions.value)
}

async function setQuestion(frame: GenerativeQnAExperience, question?: string) {
  if (question) {
    return await frame.setQuestion(question)
  } else {
    return await frame.close()
  }
}

watch(
  () => embeddingContext?.value,
  (newValue, oldValue) => {
    if (newValue && !oldValue) {
      embedQnA(newValue)
    }
  },
  { immediate: true }
)

watch(
  () => props.question,
  (newValue) => {
    if (qnaFrame.value) {
      setQuestion(qnaFrame.value, newValue)
    }
  },
  { immediate: true }
)
</script>

<template>
  <div class="v-quicksight-generative-qna" :id="containerId"></div>
</template>
