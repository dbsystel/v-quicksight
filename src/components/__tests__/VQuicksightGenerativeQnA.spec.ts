import { mount } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest'
import { nextTick, ref } from 'vue'

import { EmbeddingContextInjectionKey } from '../../symbols'
import VQuicksightGenerativeQnA from '../VQuicksightGenerativeQnA.vue'

const frameOptions = {
  container: '#custom-el',
  width: '300px',
  height: '700px',
  withIframePlaceholder: true,
  className: 'my-custom-class'
}
const contentOptions = {
  showTopicName: false,
  showPinboard: false,
  allowTopicSelection: false,
  allowFullscreen: false,
  searchPlaceholderText: ''
}

describe('VQuicksightGenerativeQnA', () => {
  function setupComponent(props: Record<string, any> = {}) {
    const qnaFrame = {
      close: vi.fn().mockResolvedValue({ success: true }),
      setQuestion: vi.fn().mockResolvedValue({ success: true })
    }
    const embeddingContext = {
      embedGenerativeQnA: vi.fn().mockResolvedValue(qnaFrame)
    }

    const component = mount(VQuicksightGenerativeQnA, {
      global: {
        provide: {
          [EmbeddingContextInjectionKey]: ref(embeddingContext)
        }
      },
      props: {
        ...props,
        url: 'https://my-embed-url.example.org',
        id: 'v-quicksight-qna-1'
      }
    })
    return { component, qnaFrame, embeddingContext }
  }

  it('renders as expected', () => {
    const { component } = setupComponent()
    expect(component.element).toMatchSnapshot()
  })

  it('should call embedQSearchBar with default parameters', async () => {
    const { embeddingContext } = setupComponent()
    await nextTick()
    expect(embeddingContext.embedGenerativeQnA).toHaveBeenCalledOnce()
    expect(embeddingContext.embedGenerativeQnA).toHaveBeenCalledWith(
      {
        className: undefined,
        container: '#v-quicksight-qna-1',
        height: '100%',
        onChange: expect.anything(),
        url: 'https://my-embed-url.example.org',
        width: '100%',
        withIframePlaceholder: false
      },
      {
        showTopicName: true,
        showPinboard: true,
        allowTopicSelection: true,
        allowFullscreen: true,
        onMessage: expect.anything()
      }
    )
  })

  it('should call embedQSearchBar with customized parameters', () => {
    const { embeddingContext } = setupComponent({ ...frameOptions, ...contentOptions })
    expect(embeddingContext.embedGenerativeQnA).toHaveBeenCalledOnce()
    expect(embeddingContext.embedGenerativeQnA).toHaveBeenCalledWith(
      {
        ...frameOptions,
        onChange: expect.anything(),
        url: 'https://my-embed-url.example.org'
      },
      {
        ...contentOptions,
        onMessage: expect.anything()
      }
    )
  })

  it('should call setQuestion when the passed question name changes', async () => {
    const { qnaFrame, component } = setupComponent()
    await nextTick() // wait until first setup is finished
    await component.setProps({ url: '', question: 'How many apples were sold?' })
    expect(qnaFrame.setQuestion).toHaveBeenCalledOnce()
    expect(qnaFrame.setQuestion).toHaveBeenCalledWith('How many apples were sold?')
  })

  it('should call close when the question changes to a nullish value', async () => {
    const { qnaFrame, component } = setupComponent({ question: 'How many apples were sold?' })
    await nextTick() // wait until first setup is finished
    await component.setProps({ url: '', question: '' })
    expect(qnaFrame.close).toHaveBeenCalled()
  })
})
