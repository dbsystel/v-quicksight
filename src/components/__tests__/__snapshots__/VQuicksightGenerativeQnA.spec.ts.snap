// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VQuicksightGenerativeQnA > renders as expected 1`] = `
<div
  class="v-quicksight-generative-qna"
  id="v-quicksight-qna-1"
/>
`;
